x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptorsIntersection.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authz;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.XContentParserUtils;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public record RoleDescriptorsIntersection(List<List<RoleDescriptor>> roleDescriptorsList) implements ToXContentObject, Writeable {
```

Comment:
I changed the type to be `Collection<Set<>>` so it is consistent. 

That said, I cannot remember why existing code uses `Collection` instead of `List`. It does not seem necessary. I used `List` for the inner type instead of `Set` in the hope to keep a predicatable order. But that was not really useful because these information is rendered as map on the user-facing output. Also we deduplicate role names when resolving role definitions. So overall `Set` is a better fit for the inner type. In summary, I think the better type would be `List<Set<>>` but this can wait for future PRs.

Reply:
Excellent and +1 on `List<Set<>>`